application/src/main/kotlin/net/corda/v5/application/persistence/PersistenceService.kt

```diff
@@ -105,79 +103,62 @@ interface PersistenceService {
      * @param namedParameters the named parameters to be set in the named query.
-     * @param postFilter the filter to be applied after named query execution.
-     * @param R the type of the results
-     * @return Cursor configured to poll data for this named query.
-     * @throws CordaPersistenceException if an error happens during query operation
-     */
-    @Suspendable
-    fun <R> query(
-        queryName: String,
-        namedParameters: Map<String, Any>,
-        postFilter: NamedQueryFilter
-    ): Cursor<R>
-
-    /**
-     * Execute a named query in a single transaction and apply post-processing to the results. Casts results to the specified type [R].
-     *
-     * @param queryName the name of the named query registered in the persistence context.
-     * @param namedParameters the named parameters to be set in the named query.
-     * @param postProcessorName the name of the post-processor that will process named query results.
-     * @param R the type of the results
-     * @return Cursor configured to poll data for this named query.
+     * @param R the type of the results.
+     * @param limit the maximum number of results to return.
+     * @return list of entities found. Empty list if none were found.
      * @throws CordaPersistenceException if an error happens during query operation
      */
     @Suspendable
     fun <R> query(
         queryName: String,
         namedParameters: Map<String, Any>,
-        postProcessorName: String
-    ): Cursor<R>
+        limit: Int,
+    ): List<R>
 
     /**
-     * Execute a named query in a single transaction and apply post-filtering and post-processing to the results. Casts results to the specified type [R].
+     * Execute a named query in a single transaction, skip results to the specified [offset] and limit the number of returned results. Casts results to the specified type [R].
      *
      * @param queryName the name of the named query registered in the persistence context.
      * @param namedParameters the named parameters to be set in the named query.
-     * @param postFilter the filter to be applied after named query execution.
-     * @param postProcessorName the name of the post-processor that will process named query results.
-     * @param R the type of the results
-     * @return Cursor configured to poll data for this named query.
+     * @param R the type of the results.
+     * @param offset the index of the first result in the query to return.
+     * @param limit the maximum number of results to return.
+     * @return list of entities found. Empty list if none were found.
      * @throws CordaPersistenceException if an error happens during query operation
      */
     @Suspendable
     fun <R> query(
         queryName: String,
         namedParameters: Map<String, Any>,
-        postFilter: NamedQueryFilter,
-        postProcessorName: String
-    ): Cursor<R>
+        offset: Int,
```

Comment:
Do we think there are ever going to be other properties to add to the query similar to `offset` and `limit`? I think having them in an object for backwards compatibility in the future would make sense without the need for tons of overloads to the `PersistenceService` (although this would be somewhat moved to the request object).

If we don't see there being many extra properties to add in the future, then what we have here is fine.

Reply:
The same crossed my mind and I was following the pattern that was there, but maybe we don't need the overloads.

If you think of the JPA API, it's quite different. There, you have query factories: `createNamedQuery()`, `createQuery()` and they return a `Query` object which is effectively a builder that exposes methods for building the query (`setParameter()`, `setMaxResults()` ... all return Query) and also a result functions (`getResultList()`, `getResultStream()` ...)
That's a quite JAVA-y API, but feels quite natural to me.

Reply:
I don't _mind_ that approach, however I do quite like the `Map<String, Any>` for kotlin usage, for Java the `setParameter` option probably still looks nicer even in newer versions of Java. Maybe we provide an option to pass in a map of parameters via `setParameters` and `setParameter` for per parameter usage. We can then remove the other APIs and just provide this one?

If we do this, then `Query` like you said in your previous comment sounds good to me.

Want more people than just me to agree though since it's drifting further away from where we were in DP1.

Also looping in @conalsmith-r3 since he created the original API. I don't think we considered doing it this way at the time so would be good to know what he thinks too.

Reply:
It would be quite nice having just one query function here returning a `Query` builder object and the user can set the parameters on it themselves, I think it would be intuitive for the user since it is similar to JPA Query object.

Reply:
Something like this?

```
interface Query<R> {
    fun setParameter(name: String, value: Any): Query<R>
    fun setParameters(parameters: Map<String, Any>): Query<R>
    fun setLimit(limit: Int): Query<R>
    fun setOffset(offset: Int): Query<R>
    fun execute(): List<R>
}
```

```
    @Suspendable
    fun <R> query(
        queryName: String
    ): Query<R>
```